### PR TITLE
Make `TcpTransport#openConnection` fully async

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is a follow-up to #35144. That commit made the underlying
connection opening process in `TcpTransport` asynchronous. However the
method still blocked on the process being complete before returning.
This commit moves the blocking to the `ConnectionManager` level. This is
another step towards the top-level `TransportService` api being async.

### CODE REVIEW

**Comment by tbrooks8** [comment]
@DaveCTurner This is another steps towards the `TransportService` api being async.

**Review by original-brownbear — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/transport/Transport.java [hunk 1]
```diff
@@ -145,6 +146,23 @@ default Object getCacheKey() {
         void close();
     }
 
+    /**
+     * This class represents a pending {@link Connection}. It provides the ability to cancel the pending
+     * connection and close the underling list of {@link TcpChannel}.
+     */
+    final class PendingConnection {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/Transport.java]: maybe we can just return a `Releasable` instead and document it's purpose?

**Review by s1monw — APPROVED** [review]
LGTM left one suggestion

**Pull request merged**